REGRESSION (iOS 17.4): Tapping on a `<datalist>` suggestion in the keyboard does not update the input's value

pxlcoder opened the pull request:
#### 8a1e52b41985bc3ea4e492dab50ec6c544243ca2
<pre>
REGRESSION (iOS 17.4): Tapping on a `&lt;datalist&gt;` suggestion in the keyboard does not update the input&apos;s value
<a href="https://bugs.webkit.org/show_bug.cgi?id=273327">https://bugs.webkit.org/show_bug.cgi?id=273327</a>
<a href="https://rdar.apple.com/124784204">rdar://124784204</a>

Reviewed by Abrar Rahman Protyasha and Wenson Hsieh.

Following BrowserEngineKit adoption in WebKit, `&lt;datalist&gt;` suggestions are now
sent to UIKit as `BETextSuggestion`s rather than `UITextSuggestion`s. However,
underlying infrastructure in UIKit only deals with `UITextSuggestion`s. As a
result, UIKit unwraps the `BETextSuggestion` into a `UITextSuggestion` upon
reception, and then converts back to a `BETextSuggestion` at insertion time.

The conversion between `BETextSuggestion` and `UITextSuggestion` breaks the
existing insertion logic, which performs a class check for `WKDataListTextSuggestion`
(a `BETextSuggestion` subclass) prior to performing insertion.

Since a class check can no longer be relied upon, insertion is now gated upon
whether the content view is aware of any active `&lt;datalist&gt;` suggestions and
whether the text of selected suggestion matches the text of a known `&lt;datalist&gt;`
suggestion.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView insertTextSuggestion:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/DataListTextSuggestionTests.mm: Added.
(-[DataListTextSuggestionInputDelegate suggestions]):
(-[DataListTextSuggestionInputDelegate setSuggestions:]):
(-[DataListTextSuggestionInputDelegate selectionDidChange:]):
(-[DataListTextSuggestionInputDelegate selectionWillChange:]):
(-[DataListTextSuggestionInputDelegate textDidChange:]):
(-[DataListTextSuggestionInputDelegate textWillChange:]):
(-[DataListTextSuggestionInputDelegate handleWebViewCredentialsSaveForWebsiteURL:user:password:passwordIsAutoGenerated:]):
(-[DataListTextSuggestionInputDelegate invalidateTextEntryContextForTextInput:]):
(-[DataListTextSuggestionInputDelegate selectionDidChangeForTextInput:]):
(-[DataListTextSuggestionInputDelegate selectionWillChangeForTextInput:]):
(-[DataListTextSuggestionInputDelegate shouldDeferEventHandlingToSystemForTextInput:context:]):
(-[DataListTextSuggestionInputDelegate textInput:deferReplaceTextActionToSystem:]):
(-[DataListTextSuggestionInputDelegate textInput:setCandidateSuggestions:]):
(-[DataListTextSuggestionTestView initWithFrame:]):
(TestWebKitAPI::TEST(DataListTextSuggestionTestView, InsertSuggestion)):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView insertTextSuggestion:]):

Canonical link: <a href="https://commits.webkit.org/278062@main">https://commits.webkit.org/278062@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0fa101681cfd7b4c5afb76a45a05c73226921b5b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49347 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28629 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52380 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/52426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/19 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51651 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34644 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26249 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/52426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51447 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26176 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42505 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/52426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23631 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43688 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7715 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45543 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44192 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54098 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24428 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20614 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47636 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/capture-with-visibility-hidden-child.html (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25700 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42712 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46628 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10849 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26539 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25424 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/TestWebKitAPI/Tests/ios/DataListTextSuggestionTests.mm:
```diff
@@ -0,0 +1,174 @@
+
+- (void)textInput:(id<BETextInput>)textInput setCandidateSuggestions:(NSArray *)suggestions
+{
+    RetainPtr uikitSuggestions = [NSMutableArray arrayWithCapacity:suggestions.count];
+    for (BETextSuggestion *beTextSuggestion in suggestions)
+        [uikitSuggestions addObject:beTextSuggestion._uikitTextSuggestion];
+
+    [self setSuggestions:uikitSuggestions.get()];
```
Does `createNSArray` accept an `NSArray *` to do the mapping?
```suggestion
    [self setSuggestions:createNSArray(suggestions, [](BETextSuggestion *beTextSuggestion) {
        return beTextSuggestion._uikitTextSuggestion;
    }];
```

aprotyas replied:
If not, then what you have is good

pxlcoder replied:
To my knowledge, it does not.

aprotyas reviewed (approved):
LGTM, small typo in the commit message

> Since a class check **c**an no longer be relied upon

pxlcoder committed:
REGRESSION (iOS 17.4): Tapping on a `<datalist>` suggestion in the keyboard does not update the input's value
https://bugs.webkit.org/show_bug.cgi?id=273327
rdar://124784204

Reviewed by Abrar Rahman Protyasha and Wenson Hsieh.

Following BrowserEngineKit adoption in WebKit, `<datalist>` suggestions are now
sent to UIKit as `BETextSuggestion`s rather than `UITextSuggestion`s. However,
underlying infrastructure in UIKit only deals with `UITextSuggestion`s. As a
result, UIKit unwraps the `BETextSuggestion` into a `UITextSuggestion` upon
reception, and then converts back to a `BETextSuggestion` at insertion time.

The conversion between `BETextSuggestion` and `UITextSuggestion` breaks the
existing insertion logic, which performs a class check for `WKDataListTextSuggestion`
(a `BETextSuggestion` subclass) prior to performing insertion.

Since a class check can no longer be relied upon, insertion is now gated upon
whether the content view is aware of any active `<datalist>` suggestions and
whether the text of selected suggestion matches the text of a known `<datalist>`
suggestion.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView insertTextSuggestion:]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/DataListTextSuggestionTests.mm: Added.
(-[DataListTextSuggestionInputDelegate suggestions]):
(-[DataListTextSuggestionInputDelegate setSuggestions:]):
(-[DataListTextSuggestionInputDelegate selectionDidChange:]):
(-[DataListTextSuggestionInputDelegate selectionWillChange:]):
(-[DataListTextSuggestionInputDelegate textDidChange:]):
(-[DataListTextSuggestionInputDelegate textWillChange:]):
(-[DataListTextSuggestionInputDelegate handleWebViewCredentialsSaveForWebsiteURL:user:password:passwordIsAutoGenerated:]):
(-[DataListTextSuggestionInputDelegate invalidateTextEntryContextForTextInput:]):
(-[DataListTextSuggestionInputDelegate selectionDidChangeForTextInput:]):
(-[DataListTextSuggestionInputDelegate selectionWillChangeForTextInput:]):
(-[DataListTextSuggestionInputDelegate shouldDeferEventHandlingToSystemForTextInput:context:]):
(-[DataListTextSuggestionInputDelegate textInput:deferReplaceTextActionToSystem:]):
(-[DataListTextSuggestionInputDelegate textInput:setCandidateSuggestions:]):
(-[DataListTextSuggestionTestView initWithFrame:]):
(TestWebKitAPI::TEST(DataListTextSuggestionTestView, InsertSuggestion)):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView insertTextSuggestion:]):

Canonical link: https://commits.webkit.org/278062@main
